### ISSUE TITLE
Slow matrix multiplication using Tensorflow 1.7.0 on a GPU

### ISSUE THREAD

**Issue by gskulkarni** [tensorflow/tensorflow#18537]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug, a feature request, or a significant problem with documentation (for small docs fixes please send a PR instead).
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Windows 10 Home (64 bit) Version 1709 OS Build 16299.371
- **TensorFlow installed from (source or binary)**:
Binary (Tensorflow GPU)
- **TensorFlow version (use command below)**:
PS C:\Users\gautam> python
Python 3.6.4 (v3.6.4:d48eceb, Dec 19 2017, 06:54:40) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> print(tf.__version__)
1.7.0
- **Python version**: 
3.6.4
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
**CUDA:**
PS C:\Users\gautam> nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2017 NVIDIA Corporation
Built on Fri_Sep__1_21:08:32_Central_Daylight_Time_2017
Cuda compilation tools, release 9.0, V9.0.176
**cuDNN:**
#define CUDNN_MAJOR 7
#define CUDNN_MINOR 0
#define CUDNN_PATCHLEVEL 5

- **GPU model and memory**:
Nvidia GTX 1050, 4GB
- **Exact command to reproduce**:

### Describe the problem
I have tensorflow-gpu installed on my Dell XPS 15 laptop running Windows 10 (environment specified above). I tried to compare matrix multiplication performance between np.matmul() and tf.matmul() for the graph mode of execution. The numpy method returns the result in about 13.5 seconds whereas the tensorflow method takes a long time and eventually fails.

### Source code / logs
I generated input matrices in the following way:
x = np.random.random((10000,10000))
y = np.random.random((10000,10000))

Here are the results:
1. np.matmul(x,y) takes about 13.5 seconds
2. The following takes a long time and eventually errors out:
with tf.Session() as sess:
    z = tf.matmul(x,y).eval()
2018-04-15 11:35:05.157088: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1344] Found device 0 with properties:
name: GeForce GTX 1050 major: 6 minor: 1 memoryClockRate(GHz): 1.493
pciBusID: 0000:01:00.0
totalMemory: 4.00GiB freeMemory: 3.30GiB
2018-04-15 11:35:05.163665: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1423] Adding visible gpu devices: 0
2018-04-15 11:35:07.338162: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:911] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-04-15 11:35:07.348029: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:917]      0
2018-04-15 11:35:07.353379: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:930] 0:   N
2018-04-15 11:35:07.363613: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1041] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 3033 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1050, pci bus id: 0000:01:00.0, compute capability: 6.1)
2018-04-15 11:36:56.906874: E T:\src\github\tensorflow\tensorflow\stream_executor\cuda\cuda_driver.cc:1110] could not synchronize on CUDA context: CUDA_ERROR_LAUNCH_FAILED ::
2018-04-15 11:36:56.906874: E T:\src\github\tensorflow\tensorflow\stream_executor\cuda\cuda_event.cc:49] Error polling for event status: failed to query event: CUDA_ERROR_LAUNCH_FAILED
2018-04-15 11:36:56.915430: F T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_event_mgr.cc:203] Unexpected Event status: 1

**Comment by Benyuel** [comment]
Instead of generating the input matrices using numpy, have you checked to see that using `tf.random_uniform` to generate the input matrices produces the same result?

**Comment by gskulkarni** [comment]
Thanks @Benyuel! That works and is about 10x as fast as the np.matmul() and I do see spikes in the GPU compute usage. In this case the random numbers were generated in the GPU, which is fine for this toy example. For real tasks, the inputs will be generated by the CPU and I'm hoping that the computations will be way more than the rate of feeding inputs - will pay more with the tf.dataset API. The CUDA error might be a bug - I don't see it for smaller input shapes.

**Comment by Benyuel** [comment]
@gskulkarni : Here's how you would generate on CPU but perform the matmul on GPU if you want to test:
```
with tf.device('/device:CPU:0'):
    x = tf.random_uniform...
    y = ...

with tf.device('/device:GPU:0'):
    z = tf.matmul(x, y)

with tf.Session() as sess:
    sess.run(z)
```

I think you may be getting the CUDA_ERROR_LAUNCH_FAILED because you ran out of memory copying `np.random.random` matrices on your GPU.  The `np.random.random` matrix is by default `np.float64` which creates a subsequent graph bigger than 2GB and errors out for me.  If I do `np.random.random((10000, 10000)).astype(np.float32)`, I can fit it in memory and compute.  Typically, it is better to use tf.float32 for speed with the price of a possible negligible loss of precision.

**Comment by reedwm** [comment]
@gskulkarni, When benchmarking, you typically want to run several warmup iterations before starting the timer. This is because TensorFlow has to do more work for the first `sess.run()` calls, such as optimizing the graph. You should also run the session call for several iterations in a loop, to reduce variance.

The CUDA_ERROR_LAUNCH_FAILED error is a separate issue than performance. Can you post a stand-alone program to reproduce? @Benyuel, I think the Tensors are too small to cause an out of memory error, because a 10000 x 10000 float64 matrix is only 80 MB. But you are right it's typically better to use float32.

**Comment by gskulkarni** [comment]
@Benyuel Thanks for the suggestion - that works for me. However, I don't see the GPU under memory pressure. @reedwm - thanks for your pointers. Here's what you can do to reproduce the error:

import numpy as np
import tensorflow as tf

x = np.random.random((10000,10000))
y = np.random.random((10000,10000))

with tf.Session() as sess:
    z = tf.matmul(x, y).eval()

**Comment by reedwm** [comment]
I cannot reproduce. @mrry, any ideas why the above code snippet is getting the error "could not synchronize on CUDA context: CUDA_ERROR_LAUNCH_FAILED"?

**Comment by mrry** [comment]
No idea. Perhaps someone with more GPU familiarity could comment.

**Comment by gskulkarni** [comment]
Perhaps related: https://github.com/tensorflow/tensorflow/issues/1450
https://github.com/tensorflow/tensorflow/issues/6783
https://github.com/tensorflow/tensorflow/issues/10852

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 105 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 125 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 140 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 155 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 170 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 185 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 200 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 215 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sushreebarsa** [comment]
@gskulkarni Is this still an issue ?
We see that you are using old version of tensorflow 1.x which is not actively supported, We recommend that you upgrade to 2.4 or later version.Please have a look at the migration guide for reference to migrate from TensorFlow 1.x to TensorFlow 2.Thank you!
